File: 프로그래머스/2/389480. 완전범죄/완전범죄.py
def solution(info, n, m):
    states = {(0, 0)}

    for a_cost, b_cost in info:
        next_states = set()

        for a, b in states:

            # A가 훔침
            if a + a_cost < n:
                next_states.add((a + a_cost, b))

            # B가 훔침
            if b + b_cost < m:
                next_states.add((a, b + b_cost))

        states = next_states

    if not states:
        return -1

    return min(a for a, b in states)

File: 프로그래머스/2/389480. 완전범죄/완전범죄.js
function solution(info, n, m) {
    let states = new Set(["0,0"]);

    for (const [aCost, bCost] of info) {
        const nextStates = new Set();

        for (const state of states) {
            const [a, b] = state.split(",").map(Number);

            // A가 훔침
            if (a + aCost < n) {
                nextStates.add(`${a + aCost},${b}`);
            }

            // B가 훔침
            if (b + bCost < m) {
                nextStates.add(`${a},${b + bCost}`);
            }
        }

        states = nextStates;
    }

    if (states.size === 0) {
        return -1;
    }

    let answer = Infinity;
    for (const state of states) {
        const [a] = state.split(",").map(Number);
        answer = Math.min(answer, a);
    }

    return answer;
}